### PR TITLE
(CDAP-8587) Reduce log frequency

### PR DESCRIPTION
**Pull request by chtyim**
This PR has two commits.

1. The first commit is a generic library introduced to create Logger with sampling through LogSampler. It's intended to be used by components in CDAP to control log frequency under different situation.
2. The second commit is mainly the fix for CDAP-8587 using LogSampler, together with some other places that I know of that can benefits from the library.

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+          break;
+        case LocationAwareLogger.ERROR_INT:
+          if (t == null) {
+            logger.error(marker, message);
+          } else {
+            logger.error(marker, message, t);
+          }
+          break;
+        default:
+          // This should happen. Just ignore it
```
Comment (albertshau) [inline cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java]: typo: "shouldn't"

File: cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+        logger.info(marker, message, arguments);
+        break;
+      case LocationAwareLogger.WARN_INT:
+        logger.warn(marker, message, arguments);
+        break;
+      case LocationAwareLogger.ERROR_INT:
+        logger.error(marker, message, arguments);
+        break;
+      default:
+        // This should happen. Just ignore it
```
Comment (albertshau) [inline cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java]: typo again

**Comment by chtyim** [comment]
Building http://builds.cask.co/browse/CDAP-RUT629

File: cdap-common/src/main/java/co/cask/cdap/common/logging/LogSamplers.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+/**
+ * Utility class to provide common implementations of {@link LogSampler}.
+ */
+public final class LogSamplers {
+
+  /**
+   * Returns a {@link LogSampler} that accepts once on every N calls.
+   */
+  public static LogSampler limit(final int count) {
```
Comment (albertshau) [inline cdap-common/src/main/java/co/cask/cdap/common/logging/LogSamplers.java]: not a big deal, but 'limit' is a bit of an odd name to me. Without looking at the javadoc or implementation, I would think it logs at most N messages. How about something like 'modulo', or 'onceEvery'

**Review by albertshau — APPROVED** [review]
minor comments, lgtm

File: cdap-common/src/main/java/co/cask/cdap/common/logging/LogSamplers.java (hunk 3 above)
Comment (chtyim) [reply cdap-common/src/main/java/co/cask/cdap/common/logging/LogSamplers.java]: I was struggling to come up with a good name for this. `onceEvery` seems like a good one.

File: cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java (hunk 1 above)
Comment (chtyim) [reply cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java]: fixed

File: cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java (hunk 2 above)
Comment (chtyim) [reply cdap-common/src/main/java/co/cask/cdap/common/logging/LocationAwareSamplingLogger.java]: fixed

**Comment by chtyim** [comment]
Build passed. Squashing and merging.

**Pull request merged**